src/components/tl/ucp/tl_ucp_context.c

```diff
@@ -13,6 +13,109 @@
 #include "schedule/ucc_schedule_pipelined.h"
 #include <limits.h>
 
+#define UCP_CHECK(function, msg, go, ctx)                                      \
+    status = function;                                                         \
+    if (UCS_OK != status) {                                                    \
+        tl_error(ctx->super.super.lib, msg ", %s", ucs_status_string(status)); \
+        ucc_status = ucs_status_to_ucc_status(status);                         \
+        goto go;                                                               \
+    }
+
+#define CHECK(test, msg, go, return_status, ctx)                               \
+    if (test) {                                                                \
+        tl_error(ctx->super.super.lib, msg);                                   \
+        ucc_status = return_status;                                            \
+        goto go;                                                               \
+    }
+
+unsigned ucc_tl_ucp_worker_progress(void *progress_arg)
+{
+    ucc_tl_ucp_context_t *ctx = (ucc_tl_ucp_context_t *)progress_arg;
+    unsigned              ret;
+
+    ret = ucp_worker_progress(ctx->worker.ucp_worker);
+
+    if (ctx->cfg.service_worker != 0) {
```

Comment:
@samnordmann now when i look at this, i think we should 2 different functions: 1 to be used when service_worker is there (like this function here) and another one with default only. We can make the decision at ctx creation and register only the required fn, thus eliminating the branch